### ISSUE TITLE
Add a filter to exclude file types when running bulp optimization

### ISSUE THREAD

**Issue by iCaspar** [wp-media/imagify-plugin#516]
We currently have a filter `imagify_auto_optimize_attachment` to disable certain file types from being optimized automatically in the upload process. Ideally we should have a filter for the bulk optimization process as well.

**Comment by tpetcoukdev** [comment]
Is there a code change that could be made to get this result. We are setting up the plugin and have over 10,000 images and 10,000 PDFs. The PDFs take considerably longer and we would like to focus on images for now. I understand that a code change would be overwritten on updates but I really do need this functionality before I hunt for a different plugin. Many thanks.

**Comment by blogtutor** [comment]
We just ran into an issue with this, on a website that sells PDF sewing patterns.  The Imagify bulk optimization flattened the layers and was causing issues for the customers.  

I've added [the code to exclude PDFs from Auto Optimization on upload](https://github.com/wp-media/imagify-helpers/tree/master/optimization/imagify-skip-pdf-auto-optimization), but if they can also be excluded from bulk optimization that would really help.

**Comment by tpetcoukdev** [comment]
I've managed to turn off bulk optimsation for PDFs by modifying /inc/functions/attachements.php

Comment out line 26 as shown below

```
function imagify_get_mime_types( $type = null ) {
	$mimes = array();

	if ( 'not-image' !== $type ) {
		$mimes = array(
			'jpg|jpeg|jpe' => 'image/jpeg',
			'png'          => 'image/png',
			'gif'          => 'image/gif',
		);
	}

	if ( 'image' !== $type ) {
                //Comment line below to prevent optimisation (including bulk) of pdfs
		//$mimes['pdf'] = 'application/pdf';
	}

	return $mimes;
}
```

This will be overwritten when the plugin is updated but is a working hotfix whilst we wait for a an update from imagify.

Hope this helps.

**Comment by mountbatt** [comment]
+1 
My pdf loose all accessibility features on upload / bulk. We need a „exclude file types“ setting in the plugin etc…

**Comment by Garconis** [comment]
Pretty please? 3 years later?

**Comment by JerissCloudCenter** [comment]
Still no news on this? This is a serious issue...

@tpetcoukdev is there no way to add this code in functions.php to avoid the code from disappearing with eveey update?

**Comment by tatof** [comment]
Bump! what is the status?  Just switched from Short pixel to this plugin. But there are some serious limits like not having the option to disable .pdf optimalisation

**Comment by JerissCloudCenter** [comment]
yes... https://github.com/wp-media/imagify-plugin/issues/516#issuecomment-1369647979 is the only solution I use in the meantime... but after every release update, you have to re-do this action

**Comment by tpetcoukdev** [comment]
> Still no news on this? This is a serious issue...
> 
> @tpetcoukdev is there no way to add this code in functions.php to avoid the code from disappearing with eveey update?

Possibly, I'll try and take some time to have a look at it this/next week. For now I can confirm that editing the code still works on this end. It would be good if Imagify could include this in an update, it really would not take much to have a checkbox in the plugin settings to "disable PDF files from bulk optimisation" - I'll even do the work if they want and they can simply test it and push it into the next update! lol

**Comment by tpetcoukdev** [comment]
Had a quick look at this, and it doesn't seem like there is a way to do this as no hooks are available on the function to allow us to override it from functions.php - there are some "hacky" workarounds but it looks like edits would still need to be made to the plugin files, which of course are lost on update.

Ref: https://stackoverflow.com/questions/33147735/override-plugin-function-in-wordpress

**Comment by rhw2008** [comment]
I have tried t install imagify-helpers-master.zip to be able to skip animated gif  images but I get an error after uploading the plugin saying :

The package could not be installed. No valid plugins were found.

Plugin installation failed.

Any help please ?

**Comment by vyskoczilova** [comment]
+1 Optimizing _images_ at the site with a couple of GB of PDFs led to twice as much storage space being used because the PDFs' originals were backed up before optimization. I don't want to optimize them at the very first.

**Comment by joffreyc-vortexsolution** [comment]
Hello!
Current solution for the plugin is to generate an error before optimization.
```
    /**
     * Stop processing optimization when it is a PDF.
     * The downside is that we have to return an error.
     *
     * @return null|WP_Error
     */
    add_filter(
      'imagify_before_optimize_size',
      function( $null, $process_class, $file ) {
        if ( 'application/pdf' === mime_content_type( $file->get_path() ) ) {
          return new WP_Error( 'no_pdf_optimization', 'PDF Optimization is disabled.' );
        }
        return $null;
      },
      10,
      3
    );

    /**
     * Prevent automatic optimization for PDF.
     *
     * @param  bool  $optimize      True to optimize, false otherwise.
     * @param  int   $attachment_id Attachment ID.
     * @return bool
     */
    function no_auto_optimize_pdf( $optimize, $attachment_id ) {
      if ( ! $optimize ) {
        return false;
      }
      return 'application/pdf' !== get_post_mime_type( $attachment_id );
    }
    add_filter( 'imagify_auto_optimize_attachment', 'no_auto_optimize_pdf', 10, 2 );
```
Wait and see if we can have an option that will prevent this in a good way.

**Comment by JerissCloudCenter** [comment]
@joffreyc-vortexsolution

Thanks for your code. This code goes in child theme > functions.php right?

So if I understand well, when the bulk optimisation runs, it will throw an error for PDFs but the bulk will not get blocked right? It will just skip and proceed to next file ?

**Comment by joffreyc-vortexsolution** [comment]
@JerissCloudCenter 

Yes, in your functions.php file.

Yes, wherever you trigger it (Bulk, media list, media item, etc.) you will see the error.

It will not stop the bulk optimization. It will say the error text for the file and go to the next file.

**Comment by JerissCloudCenter** [comment]
Ok I will try it right now. It looks like a good enough workaround for me, while waiting for the next 10 years for Imagify to implement this feature -_- Thanks!

**Comment by JerissCloudCenter** [comment]
Perfect thanks! Your code works fine, it correctly blocks the optimization of PDFs.

![image](https://github.com/wp-media/imagify-plugin/assets/87424739/22972775-2d02-4113-b2c9-e61dd90f447c)

![image](https://github.com/wp-media/imagify-plugin/assets/87424739/6945082b-f128-4d0e-9207-7eb346138e66)

**Comment by tatof** [comment]
I switched back to Shortpixel.. More advanced settings including this one, AVIF, webp etc. And also unlimited package now.

**Comment by RMCBrent** [comment]
> +1 My pdf loose all accessibility features on upload / bulk. We need a „exclude file types“ setting in the plugin etc…

I just canceled my unlimited service with imagify over this matter as the "optimization" has removed all 508/ADA/A11y features of our PDFs. 

Imagify, please prioritize this as its needed by your government and educational-system customers.

**Comment by marc-wdf** [comment]
It's incomprehensible that PDF management is still not better implemented! It's impossible to use Imagify without deleting the accessibility data included in PDFs. I'm switching to Shortpixel because of this feature.

**Comment by mrwconnected** [comment]
Seriously?  5 years later?  WTF?
